File: apps/sdkwork-appstore-mini-program/packages/sdkwork-appstore-mp-library/src/state/libraryState.ts
import type { LibraryPageResult } from "../types/libraryModels.js";

export interface LibraryState {
  readonly status: "idle" | "loading" | "ready" | "error";
  readonly lastResult?: LibraryPageResult<unknown>;
  readonly error?: string;
}

export function initialLibraryState(): LibraryState {
  return { status: "idle" };
}

export function reduceLibraryState(
  state: LibraryState,
  event:
    | { type: "load" }
    | { type: "loaded"; result: LibraryPageResult<unknown> }
    | { type: "failed"; error: string },
): LibraryState {
  switch (event.type) {
    case "load":
      return { ...state, status: "loading", error: undefined };
    case "loaded":
      return { status: "ready", lastResult: event.result };
    case "failed":
      return { ...state, status: "error", error: event.error };
  }
}

File: apps/sdkwork-appstore-flutter-mobile/packages/sdkwork_appstore_flutter_mobile_library/lib/src/state/library_state.dart
import '../models/library_models.dart';

enum LibraryStatus { idle, loading, ready, error }

class LibraryState {
  const LibraryState({
    this.status = LibraryStatus.idle,
    this.lastResult,
    this.error,
  });

  final LibraryStatus status;
  final LibraryPageResult<Object?>? lastResult;
  final String? error;

  LibraryState loading() => const LibraryState(status: LibraryStatus.loading);

  LibraryState loaded(LibraryPageResult<Object?> result) =>
      LibraryState(status: LibraryStatus.ready, lastResult: result);

  LibraryState failed(String message) =>
      LibraryState(status: LibraryStatus.error, error: message);
}
